### PR TITLE
Reduce the amount of duplicate/redundant code shared between the jpa/mongo/java/kotlin/imperative/reactive panache implementations

### PR DESCRIPTION
**Pull request by evanchooly**
This PR attempts to reduce the amount of duplicate/similar code to a much smaller subset.  I think there's still a bit more work to be done but this PR gets the large, sweeping changes out of the way. 

I included @geoand as a potential reviewer because github listed him as a recent modifier of these files.

### CODE REVIEW

**Comment by geoand** [comment]
Seems like CI is failing on `io.quarkus.hibernate.reactive.panache.test.PanacheHotReloadTestCase`

File: extensions/panache/hibernate-orm-panache-kotlin/runtime/src/main/java/io/quarkus/hibernate/orm/panache/kotlin/runtime/KotlinJpaOperations.java
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache-kotlin/runtime/src/main/java/io/quarkus/hibernate/orm/panache/kotlin/runtime/KotlinJpaOperations.java:10]: As a side note, we should really find a way for users to not use our __Operation__ classes.

File: integration-tests/hibernate-orm-panache-kotlin/src/main/kotlin/io/quarkus/it/panache/kotlin/PersonRepository.kt
Comment (loicmathieu) [inline integration-tests/hibernate-orm-panache-kotlin/src/main/kotlin/io/quarkus/it/panache/kotlin/PersonRepository.kt:10]: Tests should not be using `KotlinJpaOperations` but instead `super(query, params)`.

**Review by loicmathieu — COMMENTED** [review]
Overall it looks good to me, there is more deleted lines than added and I love it ;)

As there is a lot of changes on the deployment parts, you need @FroMage review on this one.

**Comment by evanchooly** [comment]
All the tests had been passing locally but I realized i've been using `mvn install` from the root rather than `mvn verify` and so the `integration-test` phase hadn't been running.  Using that locally, i see failures in the `maven` and `scala` integration tests but they, so far, seem unrelated.  But I'm still investigating.

File: integration-tests/hibernate-orm-panache-kotlin/src/main/kotlin/io/quarkus/it/panache/kotlin/PersonRepository.kt
Comment (evanchooly) [reply integration-tests/hibernate-orm-panache-kotlin/src/main/kotlin/io/quarkus/it/panache/kotlin/PersonRepository.kt:10]: Actually this method can probably come out.  I'd been directly implementing such methods so I had "real" bytecode to compare against.

File: extensions/panache/hibernate-orm-panache-kotlin/runtime/src/main/java/io/quarkus/hibernate/orm/panache/kotlin/runtime/KotlinJpaOperations.java
Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-kotlin/runtime/src/main/java/io/quarkus/hibernate/orm/panache/kotlin/runtime/KotlinJpaOperations.java:10]: That would be lovely, for sure.  I've tried to document them as private but that's not really going to stop anyone.

**Comment by aloubyansky** [comment]
````
/home/runner/work/quarkus/quarkus/integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/Person.java:[65,47] non-static method find(java.lang.Class<?>,java.lang.String,java.lang.Object...) cannot be referenced from a static context
````

**Comment by geoand** [comment]
The CI failures look really suspicious weird, but consistent in the various JDK jobs

**Comment by evanchooly** [comment]
I get various failures at different points in the build locally which is where I currently held up.  I fixed, it seems, the original failure.  I'm having to do a bit of reshuffling but I'm close I think but I'm having to rely on CI runs to sanity check what I've got because I can't trust the local runs overly much.

**Comment by geoand** [comment]
Completely understood

**Comment by evanchooly** [comment]
finally all tests are passing except for one where the download of something timed out.

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java:3]: wildcard imports are prohibited

File: extensions/panache/panache-common/deployment/pom.xml
Comment (loicmathieu) [inline extensions/panache/panache-common/deployment/pom.xml:28]: `panache-common` is common to Hibernate and MongoDB, it should not depends on any Hibernate modules

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheJpaEntityClassVisitor.java
Comment (loicmathieu) [inline extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheJpaEntityClassVisitor.java:19]: This class should be on `hibernate-orm-panache-common` not `panache-common`.

**Review by loicmathieu — CHANGES_REQUESTED** [review]

**Comment by loicmathieu** [comment]
@geoand should check the changes related to Spring Data extension

**Comment by geoand** [comment]
> @geoand should check the changes related to Spring Data extension

Will do soon hopefully

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java
Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java:3]: *grumble* *grumble*  between my IDE formatting and the build formatting that should've never made it in.  :)

File: extensions/panache/panache-hibernate-common/deployment/pom.xml
Comment (loicmathieu) [inline extensions/panache/panache-hibernate-common/deployment/pom.xml:23]: Could we add it as optional ?
I don't know if it's a bug library or not, but if it adds a lot of dependencies we should make it optional

**Review by loicmathieu — COMMENTED** [review]
I saw you add Hibernate Envers dependencies, could this be optional dependencies to avoid adding more libraries that needed ?

**Review by geoand — COMMENTED** [review]
This looks like a massive amount of work, really nice!

I am not the best person to review the core Panache stuff, but the Spring Data changes are pretty minimal and look perfectly OK to me

File: extensions/panache/panache-hibernate-common/deployment/pom.xml
Comment (evanchooly) [reply extensions/panache/panache-hibernate-common/deployment/pom.xml:23]: the build actually fails due to a missing module without it.  Not that things don't compile.  but literally maven stops and complains.  it's the first time I've ever seen that kind of error.  but I'll try marking it as provided and see what happens.

Comment (loicmathieu) [reply extensions/panache/panache-hibernate-common/deployment/pom.xml:23]: `optional` seems better than `provided`, but I wonder why you need it ...

Comment (evanchooly) [reply extensions/panache/panache-hibernate-common/deployment/pom.xml]: actually that looks to be vestigial. i removed it and it built locally so I think some other pom changes i made rectified the "need" for that one.

Comment (loicmathieu) [reply extensions/panache/panache-hibernate-common/deployment/pom.xml]: Cool!

**Review by loicmathieu — APPROVED** [review]
LGTM.

But you really need @FroMage to have a look at this

**Comment by evanchooly** [comment]
@FroMage?  I know it's late there but i'd like to merge this by tomorrow so if you could take a look in the morning that would be lovely.

**Comment by FroMage** [comment]
Sorry I can't review this for tomorrow. Can you wait a bit more, this is a very large PR…

**Comment by evanchooly** [comment]
Sure.  I think this area is stable enough.  the last rebase wasn't as terrible as I'd feared.

**Review by FroMage — DISMISSED** [review]
Overall this is a nice PR that simplifies things, I like when there's more deleted than added, so thanks :)
And sorry about the delay in reviewing this.
There are some questions, though, but I promise to re-review faster this time.

File: extensions/panache/hibernate-orm-panache-common/deployment/pom.xml
Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/deployment/pom.xml:38]: It *is* but i'm not sure it's necessary any longer.  some later changes in dependency trees might have resolved the error encountered here.  I will double check because I felt a bit squeamish about it, too.

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java
Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: ah.  i wasn't aware that was intended for public consumption.  i'll see what's drifted and restore it.

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheEntityClassVisitor.java
Comment (evanchooly) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheEntityClassVisitor.java:295]: I'll double check the context but reasonably comfortable with it since everything works.  but it wouldn't hurt to double check.

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheRepositoryClassVisitor.java
Comment (evanchooly) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheRepositoryClassVisitor.java:99]: I believe that stems from whether or not the ID is parameterized or not which does vary from java to kotlin and jpa to mongodb

File: extensions/panache/hibernate-orm-panache-common/deployment/pom.xml
Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/deployment/pom.xml:38]: It's there to provide access to `EnhancerConstants.PERSISTENT_FIELD_WRITER_PREFIX` in the class visitor.  If there's a more appropriate quarkus artifact (or otherwise) i'm happy to swap that out.  But it's literally for that one constant which is mildly gross but less so than copying that value in directly, I think.

Comment (FroMage) [reply extensions/panache/hibernate-orm-panache-common/deployment/pom.xml:38]: IIRC that comes from hibernate orm, not envers.

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/deployment/pom.xml:38]: ok.  i'll play with that.  thanks.

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/deployment/pom.xml]: that's exactly what I needed.  thanks.

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java
Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: The problem with these methods is that they treat `JpaOperations` as a bag of static methods but it's not any more.  So those methods would have to either change to take in the the operations reference, which would be a weird API, or just go away.  Are users actually actively using these methods?  Because I'm not seeing a non-breaking way for them to remain with everything else as it is.

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheEntityClassVisitor.java
Comment (evanchooly) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheEntityClassVisitor.java:295]: Yes.  In this case this is specific to Kotlin and how nullability is implemented.  An ID type might be `Int` but it's represented in the byte code as if it were an `int` in java.  so when mapping the entity api to the operations api, we need to map those primitive bytecode values to `Object` to resolve the generic method signature for the operation to invoke.

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheEntityClassVisitor.java
Comment (evanchooly) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheEntityClassVisitor.java:282]: This is essentially the same case as above and how nulllables are handled for "primitive" types.

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java
Comment (FroMage) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: Well, yes, those methods are used in the wild, that's the public API. I'm not sure what you mean, we currently already have several versions of this class (different packages), so as long as we keep them we know what instance of operations they delegate to, no?

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: right.  they would need to have a reference to an operations instance.  not sure how to get that in there without changing the API.  but then, we don't *want* them using the operations classes directly because those are *not* public.

Comment (FroMage) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: Just make them forward to the right `INSTANCE`, I don't see the problem.

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: well, there are two issues is see off hand (and @loicmathieu can probably weigh in here):

1. That's still a breaking change to a public API.  I personally don't see the value in this public API but that might be a discussion for another time.
1. We've been mostly working on the assumption that the operations implementation classes are internal implementation details and discouraging use outside of the generated delegations from entity/repository types.  this would officially commit us to a public, stable API.  That's not necessarily a bad thing as that API is unlikely to radically change but I do think it's a decision we should consciously make.

Adding that as a param is by far the simplest thing to do I just want to make sure we're actively choosing the consequences of it.

Comment (FroMage) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: Definitely not as a param:

```java
public class Panache {
  public static void setRollbackOnly() {
    JpaOperations.INSTANCE.setRollbackOnly();
 }
}
```

And do the same for the other existing `Panache` classes and other methods, and we're done.

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: That would create a hard dep on `hibernate-orm-panache` even for kotlin apps which could be awkward/confusing.  fwiw, the reactive form of `Panache` is a much smaller API which suggest some asymmetry in requirements/design.  Do users *really* need another helper class like this?

**Comment by evanchooly** [comment]
I'm not sure how/why that injection error started showing up.  any ideas?

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java
Comment (FroMage) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: I'm not sure what you mean about this. Currently every Panache extension has that `Panache` class. You're trying to remove it. I think you can't because people are using it. I'm not proposing to add anything new.

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: JpaOperations.INSTANCE lives in `hibernate-orm-panache` which `quarkus-hibernate-orm-panache-kotlin` does not depend on.  In order for `Panache` to function, those kotlin apps would have to import the java implementation as well.  the kotlin module does not define its `Panache` so it can't be bound to a different reference.  Now we can push down the `Panache` definitions to the language specific modules but it'd be one more point of duplication.  I have plans in the works to monitor all such duplications for divergence so maybe that's not a big deal.  But, again, it should be a conscious choice we all make.

Comment (FroMage) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: If the Kotlin version doesn't have this interface ATM, then it differs from the Java one, just move it back to the ORM module where it started. The Mongo ones have it, under their own package, so no problem there.

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: I can definitely push these types down to the specific modules.  Still trying to track down why switching from the envers dep to hibernate-orm results in that arc/injection failure.

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:18]: OK.  Simpler solution:  I can simply make those methods static on `AbstractJpaOperations` since they don't need any specialization of types between java/kotlin, e.g.  I have tests running locally now and will push as soon as they pass here.

Comment (evanchooly) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java]: everything seems to be passing now.  I've done a little bit of pom clean up where we had redundant deps.  it helped me to better track where the changes needed to be made to resolve some of the envers bits.

**Comment by evanchooly** [comment]
final thoughts?  are we good to merge?

File: extensions/panache/panache-common/deployment/pom.xml
Comment (FroMage) [inline extensions/panache/panache-common/deployment/pom.xml:38]: This is still there, but we should remove it.

**Review by FroMage — DISMISSED** [review]
Looks like you resolved the `Panache` static helper situation, so we only have to remove the dependency to the kotlin stdlib from the common module and we should be good to go.

File: extensions/panache/panache-common/deployment/pom.xml
Comment (evanchooly) [reply extensions/panache/panache-common/deployment/pom.xml:38]: I changed the scope to `compile` so it won't get pulled in unnecessarily at least.  the `javax.validation` annotation is also supported but we're in the midst of the `javax` vs `jakartaee` annotation debate so i'm not sure what to do there as such.  My gut says to retool around the EE annotations but to wait until we have an answer for those package questions.

Comment (FroMage) [reply extensions/panache/panache-common/deployment/pom.xml:38]: Isn't that the default scope? What does it change?

Comment (evanchooly) [reply extensions/panache/panache-common/deployment/pom.xml:38]: actually, you're right.  i thought it was `runtime`.  I'll switch it to `provided`.

Comment (evanchooly) [reply extensions/panache/panache-common/deployment/pom.xml:38]: Do you see a problem using the `javax.validation` bits directly?  I'd have to rework a bit of code in the class visitor but maybe it just makes sense to do that now.

Comment (FroMage) [reply extensions/panache/panache-common/deployment/pom.xml:38]: What what is using that Kotlin stdlib, though?

Comment (evanchooly) [reply extensions/panache/panache-common/deployment/pom.xml]: In this case it was only for @NotNull which came out of the kotlinc geenrated bytecode.  kotlinc will honor the java/jakarta EE annotation as well so I switched to that since we already have that in the tree.  I'll need to do a bit more work downstream but that's dependent on getting jandex to properly preserve these annotations.  In the meantime, though, this has been cleaned up.

**Comment by evanchooly** [comment]
Barring any further change requests, I plan to merge this tomorrow.

**Review by FroMage — APPROVED** [review]
OK great, all good then!

**Pull request merged**
